File: DSA/NiharikaChand_590017697/Day26Question1.java
import java.util.*;
class Day26Question1 
{
    static class Node 
    {
        int data;
        Node next;
        Node(int data) 
        {
            this.data = data;
            this.next = null;
        }
    }
    Node head;

    public void insert(int data) 
    {
        Node newNode = new Node(data);
        if (head == null) 
        {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) 
        {
            temp = temp.next;
        }
        temp.next = newNode;
    }

    public boolean swap(int k) 
    {
        int n = 0;
        Node temp = head;
        while (temp != null) 
        {
            n++;
            temp = temp.next;
        }

        k = k % n;
        if (k == 0) 
        k = n;

        if (k == n - k + 1) 
        {
             return false;
        }
        Node prev1 = null, s1 = head;
        for (int i = 1; i < k; i++) 
        {
            prev1 = s1;
            s1 = s1.next;
        }
        Node prev2 = null, s2 = head;
        for (int i = 1; i < n - k + 1; i++) 
        {
            prev2 = s2;
            s2 = s2.next;
        }
        if (prev1 != null) 
        {
            prev1.next = s2;
        } 
        else 
        {
            head = s2;
        }
        if (prev2 != null) 
        {
            prev2.next = s1;
        } 
        else 
        {
            head = s1;
        }

        Node tempNext = s1.next;
        s1.next = s2.next;
        s2.next = tempNext;
        return true;
    }

    public static void main(String args[]) 
    {
        Scanner sc = new Scanner(System.in);
        Day26Question1 list = new Day26Question1();

        System.out.print("Enter size of linkedlist: ");
        int n = sc.nextInt();
        for (int i = 0; i < n; i++) 
        {
            System.out.print("Enter element: ");
            int data = sc.nextInt();
            list.insert(data);
        }

        System.out.print("Enter value of k to swap kth node from start and end: ");
        int k = sc.nextInt();
        boolean result = list.swap(k);
        System.out.println(result);
    }
}

File: DSA/Tanvi_590016403/Day_26_Question_1.c
#include <stdio.h>
#include <stdlib.h>

// Define node structure
struct Node {
    int data;
    struct Node* next;
};

// Utility function to create new node
struct Node* newNode(int data) {
    struct Node* temp = (struct Node*)malloc(sizeof(struct Node));
    temp->data = data;
    temp->next = NULL;
    return temp;
}

// Utility function to print the list
void printList(struct Node* head) {
    while (head) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

// Append node to end of list
void append(struct Node** head_ref, int new_data) {
    struct Node* temp = newNode(new_data);
    if (*head_ref == NULL) {
        *head_ref = temp;
        return;
    }
    struct Node* last = *head_ref;
    while (last->next) last = last->next;
    last->next = temp;
}

// 1. Swap Kth Nodes from Ends
void swapKth(struct Node** head_ref, int k) {
    int n = 0;
    struct Node* temp = *head_ref;
    while (temp) {
        n++;
        temp = temp->next;
    }
    if (k > n) return;
    if (2 * k - 1 == n) return;

    struct Node *x = *head_ref, *x_prev = NULL;
    for (int i = 1; i < k; i++) {
        x_prev = x;
        x = x->next;
    }

    struct Node *y = *head_ref, *y_prev = NULL;
    for (int i = 1; i < n - k + 1; i++) {
        y_prev = y;
        y = y->next;
    }

    if (x_prev) x_prev->next = y;
    if (y_prev) y_prev->next = x;

    struct Node* temp_ptr = x->next;
    x->next = y->next;
    y->next = temp_ptr;

    if (k == 1) *head_ref = y;
    if (k == n) *head_ref = x;
}

// 2. Union of Two Linked Lists
int exists(struct Node* head, int val) {
    while (head) {
        if (head->data == val) return 1;
        head = head->next;
    }
    return 0;
}

struct Node* makeUnion(struct Node* head1, struct Node* head2) {
    struct Node* result = NULL;
    struct Node* temp = head1;

    // Add all unique elements from head1
    while (temp) {
        if (!exists(result, temp->data)) append(&result, temp->data);
        temp = temp->next;
    }

    // Add all unique elements from head2
    temp = head2;
    while (temp) {
        if (!exists(result, temp->data)) append(&result, temp->data);
        temp = temp->next;
    }

    // Sort result list using Bubble Sort (for simplicity)
    struct Node *i, *j;
    for (i = result; i && i->next; i = i->next) {
        for (j = i->next; j; j = j->next) {
            if (i->data > j->data) {
                int temp = i->data;
                i->data = j->data;
                j->data = temp;
            }
        }
    }

    return result;
}

// 3. Sort linked list of 0s, 1s and 2s
void sort012(struct Node* head) {
    int count[3] = {0, 0, 0};
    struct Node* temp = head;

    // Count 0s, 1s and 2s
    while (temp) {
        count[temp->data]++;
        temp = temp->next;
    }

    temp = head;
    int i = 0;
    while (temp) {
        if (count[i] == 0)
            ++i;
        else {
            temp->data = i;
            --count[i];
            temp = temp->next;
        }
    }
}
